### DISCUSSION TITLE
Overloading. Paramarray vs variant

### DISCUSSION THREAD

**Discussion by FullValueRider** [twinbasic/twinbasic#528] [General]
So today I once again shot myself in the  foot.

```
Public function Init(byval ipString as string, byref ipArgs as variant) as variant

and 

Public function Init(byval ipString as string, ParamArray ipArgs() as Variant) as variant
```

twinBasic is complaining at the call site that it cannot differentiate between these two forms.  However the output is

```
Public function Init(byval ipString as string, byref ipArgs as variant) as variant

vs

Public function Init(byval ipString as string, byref ParamArray ipArgs as Variant()) as variant
```

So I'm happy to go off and find an alternative route BUT the output above is comparing Variant against Variant(), so it does seem that twinbasic can differentiate between the two forms.  There is also the difference that the paramarray term appears in one form and not the other.  Is this a fundamental compiler limit (not just twinbasic) and there is no possibility of  disambiguating the two forms, or is it something on the list of things to resolve.

**Comment by bclothier** [comment]
The problem is that a `Variant` can be an array of variant, so how do you differentiate between a `Variant()` and a `Variant` that holds a `Variant()`? One way to differentiate would be with `VT_BYREF` but that is not clear from the standpoint of the language syntax. That feels like a recipe for confusion and errors so I agree that `Variant` and `Variant()` ought to be considered ambiguous and should be disallowed to avoid mixing things up.

**Comment by FullValueRider** [comment]
But my point is there is sufficient information in the method signature to allow disambiguation, one signature has paramarray the other does not, with paramarray the ipargs() is resolved as Variant(), in the other case variant is resolved as variant (which for the latter case is what I would expect. so , as a befuddled layman, 'paramarray variant', would seem to be quite different to 'Variant'.'

**Comment by retailcoder** [comment]
> But my point is there is sufficient information in the method signature to allow disambiguation, one signature has paramarray the other does not

From the perspective of the compiler trying to *resolve which overload to call*, both signatures match at the call site - the ambiguity isn't with the declarations themselves. I think failing the build is a very sensible decision here, rather than just arbitrarily picking any matching overload.

**Comment by FullValueRider** [comment]
OK, I'm remembering a previous issue I had with overloading where it was mentioned that the problem is that if you include a variant in the method signature then you can't use overloading.  

So I tried three signatures without the use of Variant

```
Public function Init(byval ipString as string, byref ipArgs as Collection) as variant
Public function Init(byval ipstring as string, byref ipArgs as ArrayList) as variant
Public function Init(byval ipString as string, byref ipArgs as Scripting.DIctionary) as variant
```
With the above I still see that twinBasic still cannot disambiguate.  Is this because Collection, ArrayList and Scripting.Dictionary are all objects?

If folks are curious as to why I'm looking at this its because I'm trying to write a string deduplication function and I want to allow ipargs to be a string, an 'item' of strings (where item supports for each), or a paramarray of strings.  I can live with substituting paramarray with an array call at the point where I provide the parameter but this doesn't avoid the variant issue.  Encapsulating the list of parameters to a specially defined class so that I use myObject(list of strings) rather then array(list of strings) led to my being surprised that I was still getting cannot disambiguate messages.

My current method has fairly complicated filter to do the disambiguation from a single method using a paramarray.  WHich also means, as a consequence, the signature really doesn't help in telling a user what types of parameter can be passed to the dedup function

```
Public Function Dedup(ByVal ipSource As String, ParamArray ipDedupArray() As Variant) As String

        If VBA.Len(ipSource) = 0 Then Return ipSource
        If VBA.IsMissing(ipDedupArray) Then ipDedupArray = Array(Char.twspace)
        If Arrays.HasNoItems(ipDedupArray) Then ipDedupArray = Array(Char.twSpace)
        
        Dim myDedupArray As Variant 
        ' DO some dodgy stuff if only one parameter is provided
        If VBA.UBound(ipDedupArray) = 0 Then
        
            select case True
                
                Case Types.IsString(ipDedupArray(0))
                    
                    If VBA.Len(ipDedupArray(0)) = 1 Then
                        
                        myDedupArray = Array(ipDedupArray(0))
                        
                        
                    Else
                        
                        If VBA.InStr(ipDedupArray(0), Char.twComma) > 0 Then
                            'Lyst is a wrapper for ArrayList
                            'I could use the 'Lyst' .ToArray method here but its not necessary as Lyst supports for Each
                            Set myDedupArray = Lyst.Deb.AddRange(VBA.Split(ipDedupArray(0), Char.twComma))
                    
                            
                        else
                            
                            myDedupArray = Array(ipDedupArray(0))
                            
                            
                        end If
                        
                    End If
                    
                Case VBA.IsArray(ipDedupArray(0))
                
                    myDedupArray = ipDedupArray(0)
                    
                
            End Select
            
        Else
            
            myDedupArray = ipDedupArray
            
            
        End If
        
        
        Dim mySource As String
        mySource = ipSource
       
        Dim myDeDup As Variant
        For Each myDeDup In myDedupArray
        
            Dim MyDedupDedup As String = myDeDup & myDeDup
            
            Do
            
                Dim myLen As Long
                myLen = Len(mySource)
                mySource = Replace(mySource, MyDedupDedup, myDeDup)
                
            Loop Until myLen = Len(mySource)
            
        Next
        
        Return mySource
        
    End Function
```
